File: baakjoon/2020/bj11568.py
import sys

n = int(input(""))
arr = list(map(int, sys.stdin.readline().split()))
dp = [-1] * n
dp[n-1] = 1
for i in range(1,n):
    index = n-i-1
    maximum = 0
    for j in range(index+1,n):
        if(arr[index] < arr[j] and dp[j] > maximum):
            maximum = dp[j]+1
    if(maximum == 0):
        dp[index] = 1
    else:
        dp[index] = maximum

print(max(dp))

File: 2020/bj11568.c
#include <stdio.h>

int arr[1001] = {};
int dp[1001] = {};

int main(){
    int i,j,n;
    scanf("%d",&n);
    for(i=0;i<n;i++){
        scanf("%d",&arr[i]);
    }

    printf("%d",lis(n));

}

int lis(int n){
    int i,j;
    int max=1;
    for(i=0;i<n;i++){
        dp[i] = 1;
        for(j=0;j<i;j++){
            if(dp[j]+1 > dp[i] && arr[j] < arr[i]){
                dp[i] = dp[j]+1;
                if(max < dp[i]){
                    max = dp[i];
                }
            }
        }
    }
    return max;
}